lighthouse-core/test/gather/driver-test.js

```diff
@@ -59,133 +49,131 @@ function createMockSendCommandFn() {
-    case 'DOM.getDocument':
-      return Promise.resolve({root: {nodeId: 249}});
-    case 'DOM.querySelector':
-      return Promise.resolve({
-        nodeId: params.selector === 'invalid' ? 0 : 231,
-      });
-    case 'DOM.querySelectorAll':
-      return Promise.resolve({
-        nodeIds: params.selector === 'invalid' ? [] : [231],
-      });
-    case 'Runtime.evaluate':
-      return Promise.resolve({result: {value: 123}});
-    case 'Runtime.getProperties':
-      return Promise.resolve({
-        result: params.objectId === 'invalid' ? [] : [{
-          name: 'test',
-          value: {
-            value: '123',
-          },
-        },
-          {
-            name: 'novalue',
-          },
-        ],
-      });
-    case 'Page.getResourceTree':
-      return Promise.resolve({frameTree: {frame: {id: 1}}});
-    case 'Page.createIsolatedWorld':
-      return Promise.resolve({executionContextId: 1});
-    case 'Network.getResponseBody':
-      return new Promise(res => setTimeout(res, MAX_WAIT_FOR_PROTOCOL + 20));
-    case 'Page.enable':
-    case 'Page.navigate':
-    case 'Page.setLifecycleEventsEnabled':
-    case 'Network.enable':
-    case 'Tracing.start':
-    case 'ServiceWorker.enable':
-    case 'ServiceWorker.disable':
-    case 'Security.enable':
-    case 'Security.disable':
-    case 'Network.setExtraHTTPHeaders':
-    case 'Network.emulateNetworkConditions':
-    case 'Emulation.setCPUThrottlingRate':
-    case 'Emulation.setScriptExecutionDisabled':
-      return Promise.resolve({});
-    case 'Tracing.end':
-      return Promise.reject(new Error('tracing not started'));
-    default:
-      throw Error(`Stub not implemented: ${command}`);
+/**
+ * In some functions we have lots of promise follow ups that get queued by protocol messages.
+ * This is a convenience method to easily advance all timers and flush all the queued microtasks.
+ */
+async function flushAllTimersAndMicrotasks() {
+  for (let i = 0; i < 1000; i++) {
+    jest.advanceTimersByTime(1);
+    await Promise.resolve();
   }
 }
 
-/* eslint-env jest */
+function createOnceStub(events) {
+  return (eventName, cb) => {
+    if (events[eventName]) {
+      // wait a tick b/c real events never fire immediately
+      setTimeout(_ => cb(events[eventName]), 0);
+      return;
+    }
+
+    throw Error(`Stub not implemented: ${eventName}`);
+  };
+}
 
 let driver;
 let connectionStub;
 
 beforeEach(() => {
   connectionStub = new Connection();
-  connectionStub.sendCommand = sendCommandOldStub;
+  connectionStub.sendCommand = cmd => {
+    throw new Error(`${cmd} not implemented`);
+  };
   driver = new Driver(connectionStub);
 });
 
 describe('.querySelector(All)', () => {
-  it('returns null when DOM.querySelector finds no node', () => {
-    return driver.querySelector('invalid').then(value => {
-      assert.equal(value, null);
-    });
+  it('returns null when DOM.querySelector finds no node', async () => {
```

Comment:
great example of how moving all the mocking local makes things much easier to follow